### PR TITLE
feat: support for getting upstream response body

### PR DESCRIPTION
**Pull request by tzssangglass**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance

- Related issues

___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

- Source branch

- Related commits and pull requests

- Target branch

### CODE REVIEW

File: runner-core/src/main/java/org/apache/apisix/plugin/runner/handler/RpcCallHandler.java [hunk 1]
```diff
@@ -106,6 +107,62 @@ protected void channelRead0(ChannelHandlerContext ctx, A6Request request) {
         }
     }
 
+    private Boolean[] fetchExtraInfo(ChannelHandlerContext ctx, PluginFilterChain chain) {
+        // fetch the nginx variables
+        Set<String> varKeys = new HashSet<>();
+        boolean requiredBody = false;
```
Comment (soulbird) [inline runner-core/src/main/java/org/apache/apisix/plugin/runner/handler/RpcCallHandler.java:113]: Better to use `requiredReqBody`.

File: tests/e2e/plugins/plugins_post_with_vars_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+			ExpectStatusRange: httpexpect.Status2xx,
+		}),
+		table.Entry("test java runner post plugin route success", tools.HttpTestCase{
+			Object:       tools.GetA6Expect(),
+			Method:       http.MethodGet,
+			Path:         "/test/java/runner/postvars",
+			ExpectStatus: http.StatusOK,
+		}),
+	)
+})
```
Comment (soulbird) [inline tests/e2e/plugins/plugins_post_with_vars_test.go:75]: The test case doesn't seem to have anything to do with the plugin. such as how you use `vars`.

Comment (tzssangglass) [reply tests/e2e/plugins/plugins_post_with_vars_test.go:75]: 1. `PostReqWithVarsFilter` change the request path from `/test/java/runner/postvars` to `/get` in `filter` by `ext-plugin-pre-req`;

2. `PostReqWithVarsFilter` **get the upstream response body(from the httpbin)** and check the `url` in the body, the `url` should be `/get`

3. the response body also one of `vars`

File: sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        body = gson.fromJson(bodyStr, body.getClass());
+        assert body.get("url").toString().endsWith((String) conf.get("rewrite_path"));
+        chain.postFilter(request, response);
+    }
+
+    @Override
+    public List<String> requiredVars() {
+        List<String> vars = new ArrayList<>();
+        vars.add("sent_http_content_type");
+        vars.add("upstream_response_length");
```
Comment (soulbird) [inline sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java:66]: What is this useful for? It is not used in the test case. What information do you want to tell the user?

File: tests/e2e/plugins/plugins_post_with_vars_test.go (hunk 2 above)
Comment (soulbird) [reply tests/e2e/plugins/plugins_post_with_vars_test.go:75]: Got it !

File: sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java (hunk 3 above)
Comment (tzssangglass) [reply sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java:66]: tells the user that they can get some variables related to upstream this way. However, I can't figure out how to verify that.

Comment (tzssangglass) [reply sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java:66]: I will continue to update and verify in another way

Comment (tzssangglass) [reply sample/src/main/java/org/apache/apisix/plugin/runner/filter/PostReqWithVarsFilter.java]: fixed

**Comment by soulbird** [comment]
Should update https://github.com/apache/apisix-java-plugin-runner/blob/main/docs/en/latest/installation-guide.md#prerequisites.

**Review by spacewander — APPROVED** [review]

**Comment by tzssangglass** [comment]
> Should update [`main`/docs/en/latest/installation-guide.md#prerequisites](https://github.com/apache/apisix-java-plugin-runner/blob/main/docs/en/latest/installation-guide.md?rgh-link-date=2022-10-09T05%3A25%3A52Z#prerequisites).

done

**Review by soulbird — APPROVED** [review]

**Pull request merged**